docs/architecture/ADR-005-liquidation.md

```diff
@@ -0,0 +1,191 @@
+
+## Decision
+
+Liquidation will require one message type (`MsgLiquidate`), one per-token parameter (`LiquidationIncentive`), and two global parameters (`MinimumCloseFactor` and `CompleteLiquidationThreshold`).
+
+There is no event type for when a borrower becomes a valid liquidation target, nor a list of valid targets stored in the module. Liquidators will have to use an off-chain tool to query their nodes periodically.
+
+## Detailed Design
+
+A function `IsLiquidationEligible(borrowerAddr)` can be created to determine if a borrower is currently exceeding their borrow limit. Any liquidation attempt against a borrower not over their limit should fail.
+
+A borrower's total borrower value (expressed in USD) can be computed from their total borrowed tokens and a price oracle.
+
+Their borrow limit is calculated similarly using borrower uToken balances, borrower collateral settings, current uToken exchange rates, and token collateral weights.
+
+### Basic Message Type
+
+To implement the liquidation functionality of the Asset Facility, one message type is required:
+
+```go
+// MsgLiquidate - a liquidator targets a specific borrower, asset type, and collateral type for complete or partial liquidation
+type MsgLiquidate struct {
+  Liquidator    sdk.AccAddress
+  Borrower      sdk.AccAddress
+  Repayment     sdk.Coin // denom + amount
+  RewardDenom   string
+}
+```
+
+Repayment's denom is the borrowed asset denom to be repaid (because the borrower may have multiple open borrows). It is always a base asset type (not a uToken).
+Its amount is the maximum amount of asset the liquidator is willing to repay. This field enables partial liquidation.
+
+RewardDenom is the collateral type which the liquidator will recieve in exchange for repaying the borrower's loan. It is always a uToken denomination.
+
+It is necessary that messages be signed by the liquidator's account. Thus the method `GetSigners` should return the `Liquidator` address for the message type above.
+
+### Partial Liquidation
+
+It is possible for the amount repaid by a liquidator to be less than the borrower's total borrow position in the target denom. Such partial liquidations can succeed without issue, reducing the borrowed amount by `Repayment.Amount` and rewarding collateral proportional to the amount repaid.
+
+Additionally, the `Repayment.Amount` found in `MsgLiquidate` functions as a maximum the liquidator is willing to repay. There are multiple factors that may reduce the actual repayment below this amount:
+
+- Borrowed amount of `Repayment.Denom` is below `Repayment.Amount`
+- Insufficient `RewardDenom` collateral to match repaid value plus liquidation incentive
+- Parameter-based maximum, e.g. if a parameter `LiquidationCap = 0.50` caps liquidated value in a single transaction at 50% of the total borrowed value.
+
+In the above scenarios, the `MsgLiquidate` should succeed with the maximum amount allowed by the conditions, rather than fail outright.
```

Comment:
We need to absolutely ensure that we emit useful events in the `MsgLiquidate` message handler. Specifically, in such a case where the full repayment is not met, the events should allude to what was actually repaid and how much. This will be useful for tracking purposes for liquidators ;-)

Reply:
Noted - when we get to events, all denominations and amounts involved in various transactions will be tracked.